provider/consul/consul_catalog.go

```diff
@@ -93,38 +100,100 @@ func (p *CatalogProvider) watchServices(stopCh <-chan struct{}) <-chan map[strin
 			default:
 			}
 
-			data, catalogMeta, err := catalog.Services(catalogOptions)
-			if err != nil {
-				log.WithError(err).Error("Failed to list services")
-				return
-			}
-
 			// Listening to changes that leads to `passing` state or degrades from it.
 			// The call is used just as a trigger for further actions
 			// (intentionally there is no interest in the received data).
-			_, healthMeta, err := health.State("passing", healthOptions)
+			_, meta, err := health.State("passing", options)
 			if err != nil {
 				log.WithError(err).Error("Failed to retrieve health checks")
 				return
 			}
 
 			// If LastIndex didn't change then it means `Get` returned
 			// because of the WaitTime and the key didn't changed.
-			sameServiceAmount := catalogOptions.WaitIndex == catalogMeta.LastIndex
-			sameServiceHealth := lastHealthIndex == healthMeta.LastIndex
-			if sameServiceAmount && sameServiceHealth {
+			if options.WaitIndex == meta.LastIndex {
 				continue
 			}
-			catalogOptions.WaitIndex = catalogMeta.LastIndex
-			lastHealthIndex = healthMeta.LastIndex
+
+			options.WaitIndex = meta.LastIndex
+
+			// The response should be unified with watchCatalogServices
+			data, _, err := catalog.Services(&api.QueryOptions{})
+			if err != nil {
+				log.Errorf("Failed to list services: %s", err)
+				return
+			}
 
 			if data != nil {
```

Comment:
I'm not seeing here any functionality that can be placed in the `else` block.
This branch statement was copied from original `watchServices` function and there wasn't any `else` block for it. 

From the source it is clear that there are two places when `nil` for `data` is possible
https://github.com/hashicorp/consul/blob/master/api/catalog.go#L140
https://github.com/hashicorp/consul/blob/master/api/catalog.go#L150
it is when we get an error which is handled at earlier phase in the code above.

Thus it seems that this `if` statement is just a precaution in case `json.NewDecoder` can bring `nil` without notifying us with an error (the actual JSON `null` in the consul response).
I doubt consul response body will be `null` within 200 OK GET response. 
Even in case there are not services at all the JSON response should include the empty map and this situation is handled already.